### ISSUE TITLE
Reproduce Table 1 of ICLR24 Paper

### ISSUE THREAD

**Issue by OctoberChang** [nilesh2797/DEXML#1]
Hi @nilesh2797 ,

Thanks for the great work of DEXML, the results are very promising.

Can you please provide the more details about how to reproduce DEXML for datasets shown in Table 1 of your ICLR paper?
For example, what's the input data format, the exact hyper-parameters, and command line usage, etc.

Very much appreciate!

**Comment by nilesh2797** [comment]
Hi @OctoberChang,

Thanks for your interest in the work. I have updated the readme to provide details on how to run the code on benchmark datasets. I am still in the process of documenting the code, so going ahead I will try to elaborate on the readme more and provide demo notebooks. We have also released the model checkpoints on huggingface (please see the updated readme for the links), hope this helps in reproducing results in the paper. Please feel free to reach out if you have any other questions/suggestions.

Thanks!
Nilesh

**Comment by OctoberChang** [comment]
@nilesh2797 ,

Thanks for your prompted update on the `README.md` and provided more detailed instructions about how to run the `DEXML` codebase. Very much appreciated! I am able to start the training job on `LF-AmazonTitles-131K` now.


Some follow-up questions regarding the hyper-parameters of `DEXML`.

Q1: The definition of `batch_size` in Table 7:  does `batch_size=1024` refers to `batch_size_per_gpu`, or the `global_batch_size` (i.e., `batch_size_per_gpu` times the number of GPUs)?
- Take `LF-AmazonTitles-131K` as an example. From Table 9, you use 4 A100 GPUs, and from Table 7, the `batch_size=1024`. Does that mean the `global_batch_size=1024x4=4096` ?

Q2: When the training `DEXML` on `LF-AmazonTitles-131K`, I saw there are some evaluation metrics logging info (see below). Are those metrics before or after the "[Reciprocal-pair Removal](http://manikvarma.org/downloads/XC/XMLRepository.html#filter)"?
```
INFO - root - 04-Apr-24 00:25:58 : Mean loss after epoch 0/100: 7.8386E+00
INFO - root - 04-Apr-24 00:26:19 :
P@1     P@3     P@5     nDCG@1  nDCG@3  nDCG@5  MRR@10  PSP@1   PSP@3   PSP@5   R@10    R@50    R@100   loss
15.93   14.35   11.04   15.93   20.73   22.83   25.69   15.15   23.04   27.33   33.77   44.56   48.68   7.8386E+00

INFO - root - 04-Apr-24 01:02:34 : Mean loss after epoch 1/100: 6.1878E+00
INFO - root - 04-Apr-24 01:38:50 : Mean loss after epoch 2/100: 5.5933E+00
INFO - root - 04-Apr-24 02:15:06 : Mean loss after epoch 3/100: 5.1147E+00
INFO - root - 04-Apr-24 02:51:22 : Mean loss after epoch 4/100: 4.7053E+00
INFO - root - 04-Apr-24 03:28:01 : Mean loss after epoch 5/100: 4.3784E+00
INFO - root - 04-Apr-24 03:28:24 :
P@1     P@3     P@5     nDCG@1  nDCG@3  nDCG@5  MRR@10  PSP@1   PSP@3   PSP@5   R@10    R@50    R@100   loss
21.73   19.36   14.92   21.73   28.18   31.06   33.44   18.57   28.83   34.56   45.27   57.12   61.36   4.3784E+00
...
INFO - root - 04-Apr-24 13:10:47 : Mean loss after epoch 21/100: 2.0527E+00
INFO - root - 04-Apr-24 13:47:04 : Mean loss after epoch 22/100: 1.9880E+00
INFO - root - 04-Apr-24 14:23:22 : Mean loss after epoch 23/100: 1.9112E+00
INFO - root - 04-Apr-24 14:59:41 : Mean loss after epoch 24/100: 1.8519E+00
INFO - root - 04-Apr-24 15:36:23 : Mean loss after epoch 25/100: 1.7886E+00
INFO - root - 04-Apr-24 15:36:46 :
P@1     P@3     P@5     nDCG@1  nDCG@3  nDCG@5  MRR@10  PSP@1   PSP@3   PSP@5   R@10    R@50    R@100   loss
24.84   22.81   17.57   24.84   32.75   36.03   37.73   20.22   33.16   39.92   51.48   62.55   66.34   1.7886E+00
```

Q3: For the training time reported in Table 9, are those measured for the full 100 epochs, or you choose an early iteration number with early stop (e.g., you pick the best model at epoch=50, then report running time for just 50 epochs, instead of the full 100 epochs)?

**Comment by nilesh2797** [comment]
Q1.
- `batch_size` is `per_gpu_batch_size` and the `lr` parameter is in a way aligned to the `per_gpu_batch_size` i.e. the code implicitly multiplies the learning rate with the number of devices used in training. For choosing a learning rate the general rule of thumb that works well for this codebase is `lr = batch_size * 10^-7` so for `batch_size = 1024` => `lr = 10^-4` 

Q2.
- If `Datasets/<dataset>/filter_labels_test.txt` file is present then it prints metrics *after* reciprocal removal otherwise it doesn't apply any filtering hence the results are *before* reciprocal removal 

Q3.
- Training times in table 9 are for full 100 epochs, usually the training does make bulk of progress in first 20-30 epochs. Also, dexml training time can improve super-linearly with number of gpus because in every batch a lot of labels are encoded so increasing batch-size reduces the number of steps in a single epoch.

Depending on available GPU resources, you can choose to run the hard negative mining dual-encoder but with a larger batch size and a high number of negatives per query which will also closely approximate dexml with full label set. See `configs/EURLex-4K/dist-de-hnm_decoupled-softmax.yaml` for a reference config file for running this kind of model training.

**Comment by OctoberChang** [comment]
Q4: Does `Datasets/<dataset>/filter_labels_train.txt` or `Datasets/<dataset>/filter_labels_test.txt` affect the training dynamics? In other words, with and without those files, does `DEXML` training end up with the same model parameters? And it is just the evaluation metric becoming different?

**Comment by nilesh2797** [comment]
Moved our further discussion to #2. Closing this issue.
